rfcs/experimental/flow_graph_wait_try_put_and_wait/README.md

```diff
@@ -0,0 +1,608 @@
+can be received from the predecessor node (explicit ``try_put_task`` call) or initially from ``try_put_and_wait``.
+
+### Queueing ``function_node``
+
+If the concurrency of the ``function_node`` is set to ``unlimited``, the node creates a task for executing the body of the node. The created task holds the reference counter on each
+``wait_context`` stored in ``metainfo`` and also wraps the ``metainfo`` object itself to broadcast it to the successors when the task is completed.
+
+If the concurrency is not ``unlimited``, the call to ``try_put_task`` tries to occupy the concurrency of the node. If the thread limit is not yet reached - behaves the same as 
+in the ``unlimited`` case. Otherwise, both input message and the metainfo are stored in the internal queue of the node. When one of the node tasks is completed, it retrieves 
+the postponed message and the corresponding metainfo from the queue and spawns a task to process it.
+
+Since the ``function_node`` guarantees that all of the elements will be retrieved from the internal queue at some time, [buffering issues](#buffering-the-metainfo) cannot take place.
+
+### Rejecting ``function_node``
+
+If the concurrency of the node is set to ``unlimited``, behaves the same as in the ``queueing`` case described above.
+
+Otherwise, if the concurrency limit of the node is reached, both message and the associated metainformations are rejected and the predecessor that called the ``try_put_task``
+is responsible on buffering both of them.
+
+If the predecessor is not a buffering node, both message and the metainfo are not processed or stored somewhere. From the waiting perspective, it means that the computations
+of the waited message are considered completed and if there are no other tasks/ buffered items corresponding to the same ``try_put_and_wait`` input in the graph, the ``try_put_and_wait``
+call is exited.
+
+When some ``function_node`` task is completed, it will try to get a buffered message and the metainfo from the predecessor by calling the ``try_get(msg, metainfo)`` method. 
+
+Since the ``function_node`` guarantees that all of the elements will be retrieved from the predecessor, [buffering issues](#buffering-the-metainfo) cannot take place
+for buffering nodes, preceding the rejecting ``function_node``.
+
+### Lightweight ``function_node``
+
+Calls to ``try_put_task`` in the lightweight node will operate on the concurrency limit of the node in the same manner as is defined by the message buffering policy -
+``queueing`` (default)  or ``rejecting``.
+
+The difference is that for lightweight nodes no tasks are created and spawned in most of the cases and the node body will be executed by the calling thread.
+Since there are no tasks, the calling thread broadcasts the output and the metainformation to the successors after completing the function.
+
+### ``continue_node``
```

Comment:
I am afraid that the overhead of metainfo for dependency graphs can be big with a straightforward implementation.

Imagine a 2D wavefront graph started with `try_put_and_wait`. Each node that receives the metainfo will send two copies of it, one for each successor. Each node in the middle of the graph will receive two piles of metainfo from its predecessors, merge those (doubling the size of the pile) and send to two its successors, This is an exponential growth pattern.

And the benefit of single-message wait for continue nodes is very doubtful, as dependency graphs are not supposed to handle several independent message flows.

I would reconsider implementation and/or applicability of this feature for continue nodes.

Reply:
Added a separate section for this issue.